### PR TITLE
Math library

### LINKED ISSUE #23: Math library
It should support Vector, Matrix and Quaternion types. BeefLang currently does not support SIMD so maybe we could wrap some existing C/C++ library if we want the performance benefits, or we can use existing [library](https://github.com/Sheepfh/tinyAlgebra)   which has MIT license or we can write new one from scratch.

### PR DESCRIPTION
**Pull request by Fusioon**
Finally made the library somehow usable. It still needs thorough testing so use with care. Some methods and operator overloads are not yet implemented but what is implemented should be enough for now. 

Closes #23

### CODE REVIEW

**Comment by RogueMacro** [comment]
@Fusioon  If I wanted to use a Vector3 then I would have to use Vector3\<float> for example?

**Comment by Fusioon** [comment]
> @Fusioon  If I wanted to use a Vector3 then I would have to use Vector3\<float> for example?

Every type can be used without specifying generic parameter. When used without generic it will use Vector3\<float> , this behaviour is defined with typealiases in Typealias.bf

File: Engine/src/Math/Matrix33.bf [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+		}
+
+		public Vector3<T> Column(int i)
+		{
+			return columns[i];
+		}
+
+		public static Self Zero => .(0,0,0,
+									 0,0,0,
+									 0,0,0);
```
Comment (RogueMacro) [inline Engine/src/Math/Matrix33.bf]: What is this indentation?

File: Engine/src/Math/Matrix33.bf [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+				tmp[1] = Vector3<T>.DotProduct(rv.Column(0), row);
+				tmp[4] = Vector3<T>.DotProduct(rv.Column(1), row);
+				tmp[7] = Vector3<T>.DotProduct(rv.Column(2), row);
+			}
+			{
+				Vector3<T> row = .(lv[2], lv[5], lv[8]);
+				tmp[2] = Vector3<T>.DotProduct(rv.Column(0), row);
+				tmp[5] = Vector3<T>.DotProduct(rv.Column(1), row);
+				tmp[8] = Vector3<T>.DotProduct(rv.Column(2), row);
+			}
```
Comment (RogueMacro) [inline Engine/src/Math/Matrix33.bf:176]: Weird indentation here too.

File: Engine/src/Math/Quaternion.bf [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+		public static Self FromAngleAxis(T angle, Vector3<T> axis)
+		{
+			let halfAngle = (T)0.5 * angle;
+			return .(axis.Normalized * (T)Math.Sin(halfAngle), (T)Math.Cos(halfAngle));
+		}
+
+		public static Vector3<T> PerpendicularVector(Vector3<T> v)
+		{
+			var axis = Vector3<T>.CrossProduct(.(1, 0, 0), v);
+			if(axis.LengthSquared > (T)0.05)
```
Comment (RogueMacro) [inline Engine/src/Math/Quaternion.bf]: Space between if and (

File: Engine/src/Math/Vector2.bf [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+		}
+
+		/// <returns>
+		/// Unsigned angle in radians between vectors
+		/// </returns>
+		public static T Angle(Self v1, Self v2)
+		{
+			let div = (v1.Length * v2.Length);
+			// div can be 0 so we need to make sure we are not dividing by zero
+			if(div == 0) return 0;
```
Comment (RogueMacro) [inline Engine/src/Math/Vector2.bf]: Space between if and (

File: Engine/src/Math/Vector4.bf [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+		}
+
+		/// <returns>
+		/// Unsigned angle in radians between vectors
+		/// </returns>
+		public static T Angle(Self v1, Self v2)
+		{
+			let div = (v1.Length * v2.Length);
+			// div can be 0 so we need to make sure we are not dividing by zero
+			if(div == 0) return 0;
```
Comment (RogueMacro) [inline Engine/src/Math/Vector4.bf]: Space between if and (

File: Engine/src/Math/Vector3.bf [hunk 6]
```diff
@@ -0,0 +1,362 @@
...
+		}
+
+		/// <returns>
+		/// Unsigned angle in radians between vectors
+		/// </returns>
+		public static T Angle(Self v1, Self v2)
+		{
+			let div = (v1.Length * v2.Length);
+			// div can be 0 so we need to make sure we are not dividing by zero
+			if(div == 0) return 0;
```
Comment (RogueMacro) [inline Engine/src/Math/Vector3.bf]: Space between if and (

**Review by RogueMacro — CHANGES_REQUESTED** [review]
I don't even understand half of the constraints

File: Engine/src/Math/Matrix33.bf (hunk 2 above)
Comment (RogueMacro) [reply Engine/src/Math/Matrix33.bf:176]: I mean some lines above this

**Review by RogueMacro — APPROVED** [review]

**Pull request merged**